src/Heatmap/HeatMapInterpolationPoints.swift

```diff
@@ -0,0 +1,399 @@
+    ///
+    /// - Parameters:
+    ///   - lat: The latitude value of the new point.
+    ///   - long: The longitude value of the new point.
+    private func append(lat: Double, long: Double) {
+        var index = 0
+        for key in data {
+            if key[0] == lat && key[1] == long {
+                data[index][2] += 1
+                return
+            }
+            index += 1
+        }
+        let temp = [lat, long, 1]
+        data.append(temp)
+    }
+    
+    // MARK: Functions that directly contribute to the creation of interpolated points
+        
+    /// A helper function that calculates the straight-line distance between two coordinates
+    ///
+    /// - Parameters:
+    ///   - lat1: The latitude value of the first point.
+    ///   - long1: The longitude value of the second point.
+    ///   - lat2: The latitude value of the second point.
+    ///   - long2: The longitude value of the second point.
+    /// - Returns: A double value representing the distance between the given points.
+    private func distance(lat1: Double, long1: Double, lat2: Double, long2: Double) -> Double {
+        return sqrt(pow(abs(lat2 - lat1), 2) + pow(abs(long2 - long1), 2))
+    }
+    
+    /// A helper function that utilizes the k-cluster algorihtm to cluster the input data points together into reasonable sets; the number of
+    /// clusters is set so that the maximum distance between the center and any point is less than a set constant value
+    ///
+    /// - Returns: A list of clusters, each of which is a list of CLLocationCoordinate2D objects.
+    private func kcluster() -> [[CLLocationCoordinate2D]] {
+        
+        // Centers contain double values representing the center of their respective clusters found
+        // in the clusters list
+        var centers = [CLLocationCoordinate2D]()
+        var clusters = [[CLLocationCoordinate2D]]()
+        
+        // Try to make as few clusters as possible; start with 1 and increment as needed
+        var numClusters = 1
+        if (data.count > 0 && data[0].count > 0) {
+            
+            // We need to keep on finding clusters until the maximum distance between the center
+            // and any point in its cluster is under a specific preset value
+            while true {
+                
+                // Set the first numClusters values in data set to be the initial cluster centers
+                for i in 0...numClusters - 1 {
+                    centers.append(CLLocationCoordinate2D(
+                        latitude: data[i][0],
+                        longitude: data[i][1])
+                    )
+                    let tempArray = [CLLocationCoordinate2D]()
+                    clusters.append(tempArray)
+                }
+                
+                // 25 iterations of updating the center and recalculating the points in that cluster
```

Comment:
This number should be declared as a constant in the file. Consider also providing this as an initializer to the class with a default value

Reply:
I kept the 25 in the comment to clarify again why 25 was chosen, but in the loop itself, it ranges from 1 to clusterIterations, where clusterIterations is default set to 25 but can be updated when the object is initiated.